pkg/generator/generator.go

```diff
@@ -0,0 +1,322 @@
+		return "", "", microerror.Mask(err)
+	}
+
+	return configmap, secrets, nil
+}
+
+// getWithPatchIfExists provides contents of filepath overwritten by patch at
+// patchFilepath. File at patchFilepath may be non-existent, resulting in pure
+// file at filepath being returned.
+func (g Generator) getWithPatchIfExists(filepath, patchFilepath string) (string, error) {
+	var err error
+
+	var base []byte
+	{
+		base, err = g.fs.ReadFile(filepath)
+		if err != nil {
+			return "", microerror.Mask(err)
+		}
+	}
+
+	// patch is not obligatory
+	if patchFilepath == "" {
+		return string(base), nil
+	}
+
+	var patch []byte
+	{
+		patch, err = g.fs.ReadFile(patchFilepath)
+		if err != nil {
+			if IsNotFound(err) {
+				return string(base), nil
+			}
+			return "", microerror.Mask(err)
+		}
+	}
+
+	result, err := applyPatch(base, patch)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+	return result, nil
+}
+
+func (g Generator) getRenderedTemplate(filepath, context string) (string, error) {
+	templateBytes, err := g.fs.ReadFile(filepath)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	result, err := g.renderTemplate(string(templateBytes), context)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	return result, nil
+}
+
+func applyPatch(base, patch []byte) (string, error) {
+	var basePathSvc *pathmodifier.Service
+	{
+		c := pathmodifier.DefaultConfig()
+		c.InputBytes = base
+		svc, err := pathmodifier.New(c)
+		if err != nil {
+			return "", microerror.Mask(err)
+		}
+		basePathSvc = svc
+	}
+
+	var patchPathSvc *pathmodifier.Service
+	{
+		c := pathmodifier.DefaultConfig()
+		c.InputBytes = patch
+		svc, err := pathmodifier.New(c)
+		if err != nil {
+			return "", microerror.Mask(err)
+		}
+		patchPathSvc = svc
+	}
+
+	patchedPaths, err := patchPathSvc.All()
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	for _, p := range patchedPaths {
+		value, err := patchPathSvc.Get(p)
+		if err != nil {
+			return "", microerror.Mask(err)
+		}
+
+		err = basePathSvc.Set(p, value)
+		if err != nil {
+			return "", microerror.Mask(err)
+		}
+	}
+
+	outputBytes, err := basePathSvc.OutputBytes()
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	return string(outputBytes), nil
+}
+
+func (g Generator) renderTemplate(templateText string, context string) (string, error) {
+	c := map[string]interface{}{}
+	err := yaml.Unmarshal([]byte(context), &c)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	funcMap := sprig.FuncMap()
+	funcMap["include"] = g.include
+
+	t := template.New("main").Funcs(funcMap)
+	err = g.addIncludeFilesToTemplate(t)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+	t = template.Must(t.Parse(templateText))
+
+	// render final template
+	out := bytes.NewBuffer([]byte{})
+	err = t.Execute(out, c)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	return out.String(), nil
+}
+
+func (g Generator) addIncludeFilesToTemplate(t *template.Template) error {
```

Comment:
Correct me if I'm wrong but it feels unnecessary. I think include could find template file dynamically as it's called. I noticed that because I wasn't sure why we add `sprig.FuncMap` in so many places.

Reply:
Cleaned up `include` logic.